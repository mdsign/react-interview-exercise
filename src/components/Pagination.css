:local(.paginationOverrides) {
    padding-left: 4px;
    margin: 5px 0 0;
}

:local(.pageCount) {
    color: #ffffff;
    font-size: 14px;
    display: inline-block;
    vertical-align: top;
    border: 1px solid #fafafa;
    border-radius: 3px;
    padding: 5px 10px;
    margin: 6px 7px 5px;
}

