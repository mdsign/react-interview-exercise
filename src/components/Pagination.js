import React, { PureComponent } from 'react';
import classnames from 'classnames';
import styles from './Pagination.css';

class Pagination extends PureComponent {
    constructor(props) {
        super(props);
        this.state = { pager: {} };
    }

    componentWillMount() {
        // set page if items array isn't empty
        if (this.props.items && this.props.items.length) {
            this.setPage(this.props.initialPage);
        }
    }

    componentDidUpdate(prevProps, prevState) {
        // reset page if items array has changed
        if (this.props.items !== prevProps.items) {
            this.setPage(this.props.initialPage);
        }
    }

    setPage(page) {
        const {items} = this.props;
        let pager = this.state.pager;

        if (page < 1 || page > pager.totalPages) {
            return;
        }

        // get new pager object for specified page
        pager = this.getPager(items.length, page);

        // get new page of items from items array
        const pageOfItems = items.slice(pager.startIndex, pager.endIndex + 1);

        // update state
        this.setState({ pager: pager });

        // call change page function in parent component
        this.props.onChangePage(pageOfItems);
    }

    getPager(totalItems, currentPage, pageSize) {
        // default to first page
        currentPage = currentPage || 1;

        // default page size is 2
        pageSize = pageSize || 2;

        // calculate total pages
        const totalPages = Math.ceil(totalItems / pageSize);

        let startPage, endPage;
        if (totalPages <= 10) {
            // less than 10 total pages so show all
            startPage = 1;
            endPage = totalPages;
        } else {
            // more than 10 total pages so calculate start and end pages
            if (currentPage <= 6) {
                startPage = 1;
                endPage = 10;
            } else if (currentPage + 4 >= totalPages) {
                startPage = totalPages - 9;
                endPage = totalPages;
            } else {
                startPage = currentPage - 5;
                endPage = currentPage + 4;
            }
        }

        // calculate start and end item indexes
        const startIndex = (currentPage - 1) * pageSize;
        const endIndex = Math.min(startIndex + pageSize - 1, totalItems - 1);

        // create an array of pages to repeat in the pager control
        const pages = [...Array((endPage + 1) - startPage).keys()].map(i => startPage + i);

        // return object with all pager properties required by the view
        return {
            totalItems: totalItems,
            currentPage: currentPage,
            pageSize: pageSize,
            totalPages: totalPages,
            startPage: startPage,
            endPage: endPage,
            startIndex: startIndex,
            endIndex: endIndex,
            pages: pages
        };
    }

    render() {
        const { pager } = this.state;

        const pageCount = pager.pages.length > 5 &&
            <div className={styles.pageCount}>
                <span>{`Showing ${pager.startIndex + 1}`}</span>
                {pager.startIndex !== pager.endIndex && <span>{`, ${pager.endIndex + 1}`}</span>}
                <span>{` of ${pager.totalItems}`}</span>
            </div>;

        return (
            <div>
                <ul className={classnames('pagination', styles.paginationOverrides)}>
                    <li className={pager.currentPage === 1 ? 'disabled' : ''}>
                        <a onClick={() => this.setPage(1)}>
                            <span aria-hidden="true">&laquo;</span>
                        </a>
                    </li>
                    <li className={pager.currentPage === 1 ? 'disabled' : ''}>
                        <a onClick={() => this.setPage(pager.currentPage - 1)}>
                            <span aria-hidden="true">&#x2039;</span>
                        </a>
                    </li>
                    {pager.pages.length <= 5 && pager.pages.map((page, index) =>
                        <li key={index} className={pager.currentPage === page ? 'active' : ''}>
                            <a onClick={() => this.setPage(page)}>{page}</a>
                        </li>
                    )}
                    <li className={pager.currentPage === pager.totalPages ? 'disabled' : ''}>
                        <a onClick={() => this.setPage(pager.currentPage + 1)}>
                            <span aria-hidden="true">&#8250;</span>
                        </a>
                    </li>
                    <li className={pager.currentPage === pager.totalPages ? 'disabled' : ''}>
                        <a onClick={() => this.setPage(pager.totalPages)}>
                            <span aria-hidden="true">&raquo;</span>
                        </a>
                    </li>
                </ul>
                {pageCount}
            </div>

        );
    }
}

export default Pagination;