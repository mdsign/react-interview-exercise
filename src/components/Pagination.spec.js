import React from 'react';
import chai, { expect } from 'chai';

import chaiEnzyme from 'chai-enzyme';
import chaiJestSnapshot from "chai-jest-snapshot";

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Pagination from './Pagination.js';

Enzyme.configure({ adapter: new Adapter() });

chai.use(chaiJestSnapshot);
chai.use(chaiEnzyme());

const data = [
    {
      name: 'Hillary Clinton',
      gender: 'female',
      starred: true
    },
    {
      name: 'Barack Obama',
      gender: 'male',
      starred: true
    }
]

describe('<Pagination /> component should', () => {
  const pagination = mount(<Pagination items={data} onChangePage={() => null} />);

  beforeEach(function() {
    chaiJestSnapshot.resetSnapshotRegistry();
    chaiJestSnapshot.setFilename(__filename + ".snap");
  });

  const props = { 
    items: data,
    onChange: () => {}, 
   };

   const pageSize = 2;

  it('exist and be a component', () => {
    expect(Pagination).to.be.a('function');
  });

  it('have at least one item to be passed to it to display the component', () => {
    expect(props.items).to.have.lengthOf.at.least(1);
  });

  it('have two items per page', () => {    
    expect(pageSize).to.equal(2);
  });

  it('return correct number of total pages', () => {
    const totalPages = props.items.length / pageSize;
    expect(totalPages).to.equal(1);
  });

  it('have pagination class name', () => {
    expect(pagination.find('.pagination')).to.have.lengthOf(1);
  });

  it('have a wrapper and navigation buttons', () => {
    expect(pagination.find('ul')).to.have.lengthOf(1);
    expect(pagination.find('li')).to.be.present;
  });

  it('render correctly', () => {
    chaiJestSnapshot.setTestName("Pagination renders correctly");
    const tree = renderer
    .create(<Pagination items={data} onChangePage={() => null} />)
    .toJSON();
    expect(tree).to.matchSnapshot();
  });
});
