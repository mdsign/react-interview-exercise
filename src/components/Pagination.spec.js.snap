// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<div>
  <ul
    className="pagination"
  >
    <li
      className="disabled"
    >
      <a
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
      </a>
    </li>
    <li
      className="disabled"
    >
      <a
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
      </a>
    </li>
    <li
      className="active"
    >
      <a
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li
      className="disabled"
    >
      <a
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
      </a>
    </li>
    <li
      className="disabled"
    >
      <a
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
      </a>
    </li>
  </ul>
</div>
`;
